pkg/csi/service/vanilla/controller.go

```diff
@@ -1137,15 +1139,157 @@ func (c *controller) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshot
+	if !isBlockVolumeSnapshotEnabled {
+		return nil, logger.LogNewErrorCode(log, codes.Unimplemented, "createSnapshot")
+	}
+
+	volumeType := prometheus.PrometheusUnknownVolumeType
+	createSnapshotInternal := func() (*csi.CreateSnapshotResponse, error) {
+		// Validate CreateSnapshotRequest
+		if err := validateVanillaCreateSnapshotRequestRequest(ctx, req); err != nil {
+			return nil, logger.LogNewErrorCodef(log, codes.Internal,
+				"validation for CreateSnapshot Request: %+v has failed. Error: %v", *req, err)
+		}
+		volumeID := req.GetSourceVolumeId()
+
+		// Check if the source volume is migrated vSphere volume
+		if strings.Contains(volumeID, ".vmdk") {
+			return nil, logger.LogNewErrorCodef(log, codes.Unimplemented,
+				"cannot snapshot migrated vSphere volume. :%q", volumeID)
+		}
+
+		// Check if volume already exists
+		volumeIds := []cnstypes.CnsVolumeId{{Id: volumeID}}
+		queryFilter := cnstypes.CnsQueryFilter{
+			VolumeIds: volumeIds,
+		}
+		queryResult, err := c.manager.VolumeManager.QueryVolume(ctx, queryFilter)
+		if err != nil {
+			return nil, logger.LogNewErrorCodef(log, codes.Internal,
+				"The volume %q to be snapshotted is not available: %v", volumeID, err)
+		}
+
+		// Check if it is a block volume
+		if queryResult.Volumes[0].VolumeType != common.BlockVolumeType {
+			return nil, logger.LogNewErrorCodef(log, codes.Unimplemented,
+				"CreateSnapshot is only supported on block volume. VolumeType: %v",
+				queryResult.Volumes[0].VolumeType)
+		}
+		volumeType = prometheus.PrometheusBlockVolumeType
+
+		// Get the snapshot size by getting the volume size, which is only valid for full backup use cases
+		var snapshotSize int64
+		if len(queryResult.Volumes) > 0 {
+			snapshotSize = queryResult.Volumes[0].BackingObjectDetails.GetCnsBackingObjectDetails().CapacityInMb
+		} else {
+			return nil, logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to get the snapshot size by querying volume: %q", volumeID)
+		}
+
+		// the returned snapshotID below is a combination of CNS VolumeID and CNS SnapshotID concatenated by the "+"
+		// sign. That is, a string of "<UUID>+<UUID>". Because, all other CNS snapshot APIs still require both
+		// VolumeID and SnapshotID as the input, while corresponding snapshot APIs in upstream CSI require SnapshotID.
+		// So, we need to bridge the gap in vSphere CSI driver and return a combined SnapshotID to CSI Snapshotter.
+		snapshotID, snapshotCreateTimePtr, err := common.CreateSnapshotUtil(ctx, c.manager, volumeID, req.Name)
+		if err != nil {
+			return nil, logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to create snapshot on volume %q: %v", volumeID, err)
+		}
+
+		snapshotCreateTimeInProto := timestamppb.New(*snapshotCreateTimePtr)
+		if err := snapshotCreateTimeInProto.CheckValid(); err != nil {
+			return nil, logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to convert creation timestamp to proto format: %v", err)
+		}
```

Comment:
```
// CheckValid returns an error if the timestamp is invalid.
// In particular, it checks whether the value represents a date that is
// in the range of 0001-01-01T00:00:00Z to 9999-12-31T23:59:59Z inclusive.
// An error is reported for a nil Timestamp.
```
Do you think this is a overkill? considering we are receiving this from CNS it's safe to assume they'll convert without issues. The nil check is also irrelevant since the time fields in snapshot is non-optional.

Reply:
Make sense. Done